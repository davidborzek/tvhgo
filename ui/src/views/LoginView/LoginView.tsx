import { useFormik } from 'formik';
import useLogin from '../../hooks/login';
import * as Yup from 'yup';

import styles from './LoginView.module.scss';
import { useTranslation } from 'react-i18next';
import Input from '../../components/Input/Input';
import Button from '../../components/Button/Button';
import FormCard from '../../components/FormCard/FormCard';
import { useRef } from 'react';
import useFormikErrorFocus from '../../hooks/formik';
import LoginFooter from '../../components/LoginFooter/LoginFooter';
import FormGroup from '../../components/Form/FormGroup/FormGroup';

const GITHUB_URL = 'https://github.com/davidborzek/tvhgo';

export default function LoginView() {
  const { t } = useTranslation();
  const { login, loading, twoFactorRequired } = useLogin();

  const usernameRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);

  const loginSchema = Yup.object().shape({
    username: Yup.string().required(t('username_required') || ''),
    password: Yup.string().required(t('password_required') || ''),
  });

  const loginFormik = useFormik({
    initialValues: {
      username: '',
      password: '',
    },
    validationSchema: loginSchema,
    onSubmit: ({ username, password }) => login(username, password),
  });

  useFormikErrorFocus(loginFormik, usernameRef, passwordRef);

  const twoFactorSchema = Yup.object().shape({
    code: Yup.string().required(t('two_factor_code_required') || ''),
  });

  const twoFactorFormik = useFormik({
    initialValues: {
      code: '',
    },
    validationSchema: twoFactorSchema,
    onSubmit: ({ code }) =>
      login(loginFormik.values.username, loginFormik.values.password, code),
  });

  const renderTwoFactorForm = () => {
    return (
      <FormCard onSubmit={twoFactorFormik.handleSubmit}>
        <FormGroup
          info={t('two_factor_auth_info')}
          direction="column"
        >
          <Input
            type="text"
            name="code"
            label={t('verification_code')}
            value={twoFactorFormik.values.code}
            onBlur={twoFactorFormik.handleBlur}
            onChange={twoFactorFormik.handleChange}
            error={
              twoFactorFormik.touched.code
                ? twoFactorFormik.errors.code
                : undefined
            }
            fullWidth
          />
          <Button
            label={t('login')}
            type="submit"
            loading={loading}
            loadingLabel={t('login_pending')}
          />
        </FormGroup>
      </FormCard>
    );
  };

  const renderLoginForm = () => {
    return (
      <FormCard onSubmit={loginFormik.handleSubmit}>
        <Input
          type="text"
          name="username"
          label={t('username')}
          value={loginFormik.values.username}
          onBlur={loginFormik.handleBlur}
          onChange={loginFormik.handleChange}
          error={
            loginFormik.touched.username
              ? loginFormik.errors.username
              : undefined
          }
          ref={usernameRef}
          fullWidth
        />
        <Input
          type="password"
          name="password"
          label={t('password')}
          value={loginFormik.values.password}
          onBlur={loginFormik.handleBlur}
          onChange={loginFormik.handleChange}
          error={
            loginFormik.touched.password
              ? loginFormik.errors.password
              : undefined
          }
          ref={passwordRef}
          fullWidth
        />
        <Button
          label={t('login')}
          type="submit"
          loading={loading}
          loadingLabel={t('login_pending')}
        />
      </FormCard>
    );
  };

  return (
    <div className={styles.Login}>
      {twoFactorRequired ? renderTwoFactorForm() : renderLoginForm()}
      <LoginFooter
        commitHash={__COMMIT_HASH__}
        githubUrl={GITHUB_URL}
        version={__VERSION__}
      />
    </div>
  );
}
