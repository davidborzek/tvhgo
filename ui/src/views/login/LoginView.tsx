import { useRef } from 'react';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import { useTranslation } from 'react-i18next';

import useLogin from '@/hooks/login';
import Input from '@/components/common/input/Input';
import Button from '@/components/common/button/Button';
import LoginCard from '@/components/login/card/LoginCard';
import useFormikErrorFocus from '@/hooks/formik';
import LoginFooter from '@/components/login/footer/LoginFooter';
import FormGroup from '@/components/common/form/FormGroup/FormGroup';

import styles from './LoginView.module.scss';

const GITHUB_URL = 'https://github.com/davidborzek/tvhgo';

export default function LoginView() {
  const { t } = useTranslation();
  const { login, loading, twoFactorRequired } = useLogin();

  const usernameRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);

  const loginSchema = Yup.object().shape({
    username: Yup.string().required(t('username_required')),
    password: Yup.string().required(t('password_required')),
  });

  const loginFormik = useFormik({
    initialValues: {
      username: '',
      password: '',
    },
    validationSchema: loginSchema,
    onSubmit: ({ username, password }) => login(username, password),
  });

  useFormikErrorFocus(loginFormik, usernameRef, passwordRef);

  const twoFactorSchema = Yup.object().shape({
    code: Yup.string().required(t('two_factor_code_required') || ''),
  });

  const twoFactorFormik = useFormik({
    initialValues: {
      code: '',
    },
    validationSchema: twoFactorSchema,
    onSubmit: ({ code }) =>
      login(loginFormik.values.username, loginFormik.values.password, code),
  });

  const renderTwoFactorForm = () => {
    return (
      <LoginCard onSubmit={twoFactorFormik.handleSubmit}>
        <FormGroup info={t('two_factor_auth_info')} direction="column">
          <Input
            type="text"
            name="code"
            label={t('verification_code')}
            value={twoFactorFormik.values.code}
            onBlur={twoFactorFormik.handleBlur}
            onChange={twoFactorFormik.handleChange}
            error={
              twoFactorFormik.touched.code
                ? twoFactorFormik.errors.code
                : undefined
            }
            fullWidth
          />
          <Button
            label={t('login')}
            type="submit"
            loading={loading}
            loadingLabel={t('login_pending')}
          />
        </FormGroup>
      </LoginCard>
    );
  };

  const renderLoginForm = () => {
    return (
      <LoginCard onSubmit={loginFormik.handleSubmit}>
        <Input
          type="text"
          name="username"
          label={t('username')}
          value={loginFormik.values.username}
          onBlur={loginFormik.handleBlur}
          onChange={loginFormik.handleChange}
          error={
            loginFormik.touched.username
              ? loginFormik.errors.username
              : undefined
          }
          ref={usernameRef}
          fullWidth
        />
        <Input
          type="password"
          name="password"
          label={t('password')}
          value={loginFormik.values.password}
          onBlur={loginFormik.handleBlur}
          onChange={loginFormik.handleChange}
          error={
            loginFormik.touched.password
              ? loginFormik.errors.password
              : undefined
          }
          ref={passwordRef}
          fullWidth
        />
        <Button
          label={t('login')}
          type="submit"
          loading={loading}
          loadingLabel={t('login_pending')}
        />
      </LoginCard>
    );
  };

  return (
    <div className={styles.Login}>
      {twoFactorRequired ? renderTwoFactorForm() : renderLoginForm()}
      <LoginFooter
        commitHash={__COMMIT_HASH__}
        githubUrl={GITHUB_URL}
        version={__VERSION__}
      />
    </div>
  );
}
