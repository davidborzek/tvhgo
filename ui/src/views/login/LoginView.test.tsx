import { cleanup, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { afterEach, expect, vi, test, describe } from 'vitest';
import LoginView from './LoginView';
import useLogin from '@/hooks/login';
import userEvent from '@testing-library/user-event';

vi.mock('@/hooks/login');

afterEach(() => {
  vi.resetAllMocks();
  cleanup();
});

describe('login', () => {
  test('should render', () => {
    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: () => {},
      twoFactorRequired: false,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    expect(document.asFragment()).toMatchSnapshot();
  });

  test('should render loading button', () => {
    vi.mocked(useLogin).mockReturnValue({
      loading: true,
      login: () => {},
      twoFactorRequired: false,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    expect(document.getByText('login_pending')).toBeInTheDocument();
  });

  test('should show errors when no input is present', async () => {
    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: () => {},
      twoFactorRequired: false,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    await userEvent.click(document.getByText('login'));
    await document.findAllByText('username_required');

    expect(document.getByText('username_required')).toBeInTheDocument();
    expect(document.getByText('password_required')).toBeInTheDocument();
  });

  test('should show username_required when username is missing', async () => {
    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: () => {},
      twoFactorRequired: false,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    const passwordInput = document.container.querySelector(
      'input[name="password"]'
    ) as Element;

    await userEvent.type(passwordInput, 'password');

    await userEvent.click(document.getByText('login'));
    await document.findAllByText('username_required');

    expect(document.getByText('username_required')).toBeInTheDocument();
    expect(document.queryByText('password_required')).not.toBeInTheDocument();
  });

  test('should show password_required when password is missing', async () => {
    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: () => {},
      twoFactorRequired: false,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    const usernameInput = document.container.querySelector(
      'input[name="username"]'
    ) as Element;

    await userEvent.type(usernameInput, 'username');

    await userEvent.click(document.getByText('login'));
    await document.findAllByText('password_required');

    expect(document.getByText('password_required')).toBeInTheDocument();
    expect(document.queryByText('username_required')).not.toBeInTheDocument();
  });

  test('should login', async () => {
    const loginMock = vi.fn();
    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: loginMock,
      twoFactorRequired: false,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    const usernameInput = document.container.querySelector(
      'input[name="username"]'
    ) as Element;
    const passwordInput = document.container.querySelector(
      'input[name="password"]'
    ) as Element;

    const username = 'myUsername';
    const password = 'myPassword';

    await userEvent.type(usernameInput, username);
    await userEvent.type(passwordInput, password);

    expect(usernameInput).toHaveValue(username);
    expect(passwordInput).toHaveValue(password);

    await userEvent.click(document.getByText('login'));

    expect(loginMock).toHaveBeenNthCalledWith(1, username, password);
  });
});

describe('login with totp', () => {
  test('should render two factor auth', () => {
    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: () => {},
      twoFactorRequired: true,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    expect(document.asFragment()).toMatchSnapshot();
  });

  test('should render loading button', () => {
    vi.mocked(useLogin).mockReturnValue({
      loading: true,
      login: () => {},
      twoFactorRequired: true,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    expect(document.getByText('login_pending')).toBeInTheDocument();
  });

  test('should show two_factor_code_required when code is missing', async () => {
    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: () => {},
      twoFactorRequired: true,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    await userEvent.click(document.getByText('login'));
    await document.findAllByText('two_factor_code_required');

    expect(document.getByText('two_factor_code_required')).toBeInTheDocument();
  });

  test('should login with totp', async () => {
    const loginMock = vi.fn();
    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: loginMock,
      twoFactorRequired: false,
    });

    const document = render(<LoginView />, { wrapper: MemoryRouter });

    const usernameInput = document.container.querySelector(
      'input[name="username"]'
    ) as Element;
    const passwordInput = document.container.querySelector(
      'input[name="password"]'
    ) as Element;

    const username = 'myUsername';
    const password = 'myPassword';
    const code = '123456';

    await userEvent.type(usernameInput, username);
    await userEvent.type(passwordInput, password);

    expect(usernameInput).toHaveValue(username);
    expect(passwordInput).toHaveValue(password);

    vi.mocked(useLogin).mockReturnValue({
      loading: false,
      login: loginMock,
      twoFactorRequired: true,
    });

    document.rerender(<LoginView />);

    const codeInput = document.container.querySelector(
      'input[name="code"]'
    ) as Element;

    await userEvent.type(codeInput, code);

    await userEvent.click(document.getByText('login'));

    expect(loginMock).toHaveBeenNthCalledWith(1, username, password, code);
  });
});
