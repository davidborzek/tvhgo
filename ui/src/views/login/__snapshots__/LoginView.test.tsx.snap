// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`login > should render 1`] = `
<DocumentFragment>
  <div
    class="_Login_d6adf2"
  >
    <div
      class="_card_fb3893"
    >
      <div
        class="_imageHeader_fb3893"
      >
        <img
          class="_image_fb3893"
          src="/img/tvhgo.png"
        />
      </div>
      <form
        class="_form_fb3893"
      >
        <div
          class="_inputContainer_ae92e9"
          style="width: 100%;"
        >
          <label
            class="_inputLabel_ae92e9"
            for="username"
          >
            username
          </label>
          <div
            class="_innerContainer_ae92e9"
          >
            <input
              class="_input_ae92e9"
              name="username"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="_inputContainer_ae92e9"
          style="width: 100%;"
        >
          <label
            class="_inputLabel_ae92e9"
            for="password"
          >
            password
          </label>
          <div
            class="_innerContainer_ae92e9"
          >
            <input
              class="_input_ae92e9"
              name="password"
              type="password"
              value=""
            />
          </div>
        </div>
        <button
          class="_button_a7f9df"
          type="submit"
        >
          login
        </button>
      </form>
    </div>
    <footer
      class="_footer_223d5a"
    >
      <span>
        made_with
      </span>
      <span
        class="_sep_223d5a"
      >
         | 
      </span>
      <span>
        <a
          class="_link_223d5a"
          href="https://github.com/davidborzek/tvhgo"
          rel="noopener noreferrer"
          target="_blank"
        >
          Open Source
        </a>
      </span>
      <span
        class="_sep_223d5a"
      >
         | 
      </span>
      <span>
        <a
          class="_link_223d5a"
          href="https://github.com/davidborzek/tvhgo/commit/local"
          rel="noopener noreferrer"
          target="_blank"
        >
          local (local)
        </a>
      </span>
    </footer>
  </div>
</DocumentFragment>
`;

exports[`login with totp > should render two factor auth 1`] = `
<DocumentFragment>
  <div
    class="_Login_d6adf2"
  >
    <div
      class="_card_fb3893"
    >
      <div
        class="_imageHeader_fb3893"
      >
        <img
          class="_image_fb3893"
          src="/img/tvhgo.png"
        />
      </div>
      <form
        class="_form_fb3893"
      >
        <div
          class="_FormGroup_d7b0a1"
        >
          <div
            class="_content_d7b0a1 _column_d7b0a1"
          >
            <div
              class="_inputContainer_ae92e9"
              style="width: 100%;"
            >
              <label
                class="_inputLabel_ae92e9"
                for="code"
              >
                verification_code
              </label>
              <div
                class="_innerContainer_ae92e9"
              >
                <input
                  class="_input_ae92e9"
                  name="code"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <button
              class="_button_a7f9df"
              type="submit"
            >
              login
            </button>
          </div>
          <span
            class="_info_d7b0a1"
          >
            two_factor_auth_info
          </span>
        </div>
      </form>
    </div>
    <footer
      class="_footer_223d5a"
    >
      <span>
        made_with
      </span>
      <span
        class="_sep_223d5a"
      >
         | 
      </span>
      <span>
        <a
          class="_link_223d5a"
          href="https://github.com/davidborzek/tvhgo"
          rel="noopener noreferrer"
          target="_blank"
        >
          Open Source
        </a>
      </span>
      <span
        class="_sep_223d5a"
      >
         | 
      </span>
      <span>
        <a
          class="_link_223d5a"
          href="https://github.com/davidborzek/tvhgo/commit/local"
          rel="noopener noreferrer"
          target="_blank"
        >
          local (local)
        </a>
      </span>
    </footer>
  </div>
</DocumentFragment>
`;
